src/Layout.c

```diff
@@ -1133,191 +1282,449 @@ static void layoutNodeImpl(css_node_t *node, float parentMaxWidth, float parentM
+
+static const char* spacer = "                                                            ";
+
+static const char* getSpacer(unsigned long level) {
+  unsigned long spacerLen = strlen(spacer);
+  if (level > spacerLen) {
+    level = spacerLen;
+  }
+  return &spacer[spacerLen - level];
+}
+
+static const char* getModeName(css_measure_mode_t mode, bool performLayout) {
+  const char* kMeasureModeNames[CSS_MEASURE_MODE_COUNT] = {
+    "UNDEFINED",
+    "EXACTLY",
+    "AT_MOST"
+  };
+  const char* kLayoutModeNames[CSS_MEASURE_MODE_COUNT] = {
+    "LAY_UNDEFINED",
+    "LAY_EXACTLY",
+    "LAY_AT_MOST"
+  };
+  
+  if (mode >= CSS_MEASURE_MODE_COUNT) {
+    return "";
+  }
+  
+  return performLayout? kLayoutModeNames[mode] : kMeasureModeNames[mode];
+}
+
+static bool canUseCachedMeasurement(float availableWidth, float availableHeight,
+  float marginRow, float marginColumn,
+  css_measure_mode_t widthMeasureMode, css_measure_mode_t heightMeasureMode,
+  css_cached_measurement_t cachedLayout) {
+
+  // Is it an exact match?
+  if (eq(cachedLayout.available_width, availableWidth) &&
+      eq(cachedLayout.available_height, availableHeight) &&
+      cachedLayout.width_measure_mode == widthMeasureMode &&
+      cachedLayout.height_measure_mode == heightMeasureMode) {
+    return true;
+  }
+  
+  // If the width is an exact match, try a fuzzy match on the height.
+  if (cachedLayout.width_measure_mode == widthMeasureMode &&
+      eq(cachedLayout.available_width, availableWidth) &&
+      heightMeasureMode == CSS_MEASURE_MODE_EXACTLY &&
+      eq(availableHeight - marginColumn, cachedLayout.computed_height)) {
+    return true;
+  }
+  
+  // If the height is an exact match, try a fuzzy match on the width.
+  if (cachedLayout.height_measure_mode == heightMeasureMode &&
+      eq(cachedLayout.available_height, availableHeight) &&
+      widthMeasureMode == CSS_MEASURE_MODE_EXACTLY &&
+      eq(availableWidth - marginRow, cachedLayout.computed_width)) {
+    return true;
+  }
+
+  return false;
+}
+
+//
+// This is a wrapper around the layoutNodeImpl function. It determines
+// whether the layout request is redundant and can be skipped.
+//
+// Parameters:
+//  Input parameters are the same as layoutNodeImpl (see above)
+//  Return parameter is true if layout was performed, false if skipped
+//
+bool layoutNodeInternal(css_node_t* node, float availableWidth, float availableHeight,
+    css_direction_t parentDirection, css_measure_mode_t widthMeasureMode, css_measure_mode_t heightMeasureMode, bool performLayout, char* reason) {
+  css_layout_t* layout = &node->layout;
+  
+  gDepth++;
+
+  bool needToVisitNode = (node->is_dirty(node->context) && layout->generation_count != gCurrentGenerationCount) ||
+    layout->last_parent_direction != parentDirection;
+
+  if (needToVisitNode) {
+    // Invalidate the cached results.
+    layout->next_cached_measurements_index = 0;
+    layout->cached_layout.width_measure_mode = (css_measure_mode_t)-1;
```

Comment:
Thanks for the feedback.

It means that `width_measure_mode` is uninitialized. The idea is that while a cache entry's `width_measure_mode` has this value, `layoutNodeInternal` will never return that cache entry because the `widthMeasureMode` given to `layoutNodeInternal` will never match `-1`.

This pattern was used for `layout.last_direction` prior to this PR: https://github.com/facebook/css-layout/blob/b0d00ad33850d83450139d994bded89d20ddac32/src/Layout.c#L80

If we were to create an explicit enum value instead of using `-1`, it might give people the idea that it is okay to call `layoutNode` with that enum value for `widthMeasureMode` or `heightMeasureMode` which is not the case. This could be mitigated with documentation.

That's the rationale. I'm not sure whether or not having an explicit enum value would result in more idiomatic C code.